internal/pkg/vm/internal/vmcontext/vmcontext.go

```diff
@@ -1,718 +1,640 @@
+		panic("unreachable: negative funds transfer not allowed")
+	}
 
-func computeActorAddress(creator address.Address, nonce uint64) (address.Address, error) {
-	buf := new(bytes.Buffer)
+	ctx := context.Background()
 
-	if _, err := buf.Write(creator.Bytes()); err != nil {
-		return address.Undef, err
+	// retrieve debit account
+	fromActor, err := vm.state.GetActor(ctx, debitFrom)
+	if err != nil {
+		panic(fmt.Errorf("unreachable: debit account not found. %s", err))
 	}
 
-	if err := binary.Write(buf, binary.BigEndian, nonce); err != nil {
-		return address.Undef, err
+	// check that account has enough balance for transfer
+	if fromActor.Balance.LessThan(amount) {
+		panic("unreachable: insufficient balance on debit account")
 	}
 
-	return address.NewActorAddress(buf.Bytes())
-}
+	// retrieve credit account
+	toActor, err := vm.state.GetActor(ctx, creditTo)
+	if err != nil {
+		panic(fmt.Errorf("unreachable: credit account not found. %s", err))
+	}
 
-//
-// internal methods not exposed to actors
-//
+	// move funds
+	fromActor.Balance = fromActor.Balance.Sub(amount)
+	toActor.Balance = toActor.Balance.Add(amount)
 
-// AllowSideEffects determines wether or not the actor code is allowed to produce side-effects.
-//
-// At this time, any `Send` to the same or another actor is considered a side-effect.
-func (ctx *VMContext) AllowSideEffects(allow bool) {
-	ctx.allowSideEffects = allow
+	return
 }
 
-// ExtendedRuntime has a few extra methods on top of what is exposed to the actors.
-type ExtendedRuntime interface {
-	runtime.Runtime
-	LegacyMessage() *types.UnsignedMessage
-	From() *actor.Actor
-	To() *actor.Actor
-	Actors() ExecutableActorLookup
+func (vm *VM) getActorImpl(code cid.Cid) dispatch.ExecutableActor {
+	actorImpl, err := vm.actorImpls.GetActorImpl(code, vm.currentEpoch)
+	if err != nil {
+		runtime.Abort(exitcode.ActorCodeNotFound)
+	}
+	return actorImpl
 }
 
-var _ ExtendedRuntime = (*VMContext)(nil)
+//
+// implement runtime.Runtime for VM
+//
 
-// Actors returns the executable actors lookup table.
-func (ctx *VMContext) Actors() ExecutableActorLookup {
-	return ctx.actors
-}
+var _ runtime.Runtime = (*VM)(nil)
 
-// From returns the actor the message originated from.
-func (ctx *VMContext) From() *actor.Actor {
-	return ctx.from
+// CurrentEpoch implements runtime.Runtime.
+func (vm *VM) CurrentEpoch() types.BlockHeight {
+	return vm.currentEpoch
 }
 
-// To returns the actor the message is intended for.
-func (ctx *VMContext) To() *actor.Actor {
-	return ctx.to
+// Randomness implements runtime.Runtime.
+func (vm *VM) Randomness(epoch types.BlockHeight) runtime.Randomness {
+	return vm.rnd.Randomness(epoch)
 }
 
-var _ ExportContext = (*VMContext)(nil)
+// Storage implements runtime.Runtime.
+func (vm *VM) Storage() runtime.Storage {
+	return actorStorage{inner: vm.store}
+}
 
-// Params implements ExportContext.
-func (ctx *VMContext) Params() []byte {
-	return ctx.message.Params
+// LegacyStorage implements runtime.Runtime.
```

Comment:
Do we still need this?

Reply:
Yes, didn't want to peel off that layer of the onion on the same PR.

Honestly, it would be faster to hook the new actor code so we can delete all the current actors, then delete the rest of the legacy methods.